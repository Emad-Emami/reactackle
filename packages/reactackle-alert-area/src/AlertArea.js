import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { noop, registerDefaultComponentTheme } from 'reactackle-core';
import { Alert } from './Alert/Alert';
import componentTheme from './styles/theme';
import { AlertsWrapperStyled } from './styles/AlertsWrapperStyled';

registerDefaultComponentTheme('alertsArea', componentTheme);

const propTypes = {
  /*
   * Defines timeout for Alert close event
   */
  defaultTimeout: PropTypes.number,
  /*
   * Array of Alerts
   */
  alerts: PropTypes.arrayOf(Alert),
  /*
   * Specify function to be called on Close button click
   */
  onClose: PropTypes.func,
  /*
   * Speed of exit animation
   */
  animationDuration: PropTypes.number,
  /*
   * Horizontal position of Alert area relatively ro the window
   */
  positionX: PropTypes.oneOf(['left', 'right']),
  /*
   * Vertical position of Alert area relatively ro the window
   */
  positionY: PropTypes.oneOf(['top', 'bottom']),
};
const defaultProps = {
  defaultTimeout: 3000,
  alerts: [],
  onClose: noop,
  icon: null,
  animationDuration: 300,
  positionX: 'right',
  positionY: 'bottom',
};

export default class AlertArea extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentAlert: 0,
      queue: [],
      closeAnimation: false,
    };
    this._timeoutId = null;
    this._handleAlertClose = this._handleAlertClose.bind(this);
    this._closeCurrentAlert = this._closeCurrentAlert.bind(this);
  }
  
  componentDidMount() {
    this._setupTimeout();
  }
  
  componentWillReceiveProps(nextProps) {
    if (nextProps.alerts !== this.props.alerts) {
      this.setState({
        currentAlert: 0,
        queue: [],
        closeAnimation: false,
      }, () => {
        this._setupTimeout();
      });
    }
  }
  
  _handleAlertClose() {
    this.setState({
      closeAnimation: true,
    });
    this._timeoutId = setTimeout(
      this._closeCurrentAlert,
      this.props.animationDuration,
    );
  }
  
  _closeCurrentAlert() {
    const alert = this._getAlert();
    if (alert) {
      this.setState(prevState => ({
        currentAlert: prevState.currentAlert + 1,
        closeAnimation: false,
      }), () => {
        this.props.onClose(alert.id);
        this._setupTimeout();
      });
    }
  }
  
  _setupTimeout() {
    this._timeoutId = null;
    const alert = this._getAlert();
    if (alert) {
      const timeout = (alert.timeout === 0 || alert.timeout)
        ? +alert.timeout
        : this.props.defaultTimeout;
      if (timeout) {
        this._timeoutId = timeout > this.props.animationDuration
          ? setTimeout(
            this._handleAlertClose,
            timeout - this.props.animationDuration,
          )
          : setTimeout(this._closeCurrentAlert, timeout);
      }
    }
  }
  
  _getAlert() {
    const { currentAlert, queue } = this.state;
    const queueIndex = currentAlert - this.props.alerts.length;
    if (currentAlert < this.props.alerts.length)
      return this.props.alerts[currentAlert];
    else if (queueIndex < queue.length)
      return queue[queueIndex];
    return null;
  }
  
  _renderCurrentAlert() {
    const { closeAnimation } = this.state;
    const alert = this._getAlert();
    if (alert) {
      return (
        <Alert
          id={alert.id}
          closeAnimation={closeAnimation}
          onClose={this._handleAlertClose}
          {...alert}
        >
          {alert.content}
        </Alert>
      );
    }
    return null;
  }
  
  addToQueue(alert) {
    this.setState({
      queue: [
        ...this.state.queue,
        alert,
      ],
    }, () => {
      if (!this._timeoutId)
        this._setupTimeout();
    });
  }
  
  render() {
    const currentAlert = this._renderCurrentAlert();
    return (
      <AlertsWrapperStyled
        positionX={this.props.positionX}
        positionY={this.props.positionY}
      >
        {currentAlert}
      </AlertsWrapperStyled>
    );
  }
}

AlertArea.propTypes = propTypes;
AlertArea.defaultProps = defaultProps;
AlertArea.displayName = 'AlertArea';
