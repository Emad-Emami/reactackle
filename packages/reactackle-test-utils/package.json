{
  "name": "reactackle-test-utils",
  "version": "1.0.0-beta.7",
  "description": "Test utils of the Reactackle ui library",
  "scripts": {
    "prepublishOnly": "rimraf ./dist && npx babel ./src --out-dir ./dist"
  },
  "author": "braincrumbs",
  "license": "MIT",
  "keywords": [
    "react",
    "components",
    "reactackle"
  ],
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "dependencies": {
    "enzyme": "^3.1.0",
    "enzyme-to-json": "^3.1.4",
    "react-test-renderer": "^16.0.0"
  },
  "peerDependencies": {
    "react": "^16.0.0"
  },
  "devDependencies": {
    "babel-cli": "^6.24.1",
    "rimraf": "^2.6.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/bcrumbs/reactackle.git"
  },
  "bugs": "https://github.com/bcrumbs/reactackle/issues",
  "homepage": "https://reactackle.braincrumbs.io/"
}
