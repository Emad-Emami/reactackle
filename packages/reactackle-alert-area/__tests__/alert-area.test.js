
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { findByType } from 'reactackle-test-utils';
import AlertArea from '../src/AlertArea';

jest.mock('react-dom');
jest.useFakeTimers();

describe('<AlertArea/>', () => {
  it('renders correctly with default props', () => {
    const tree = renderer.create(
      <AlertArea />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if prop alerts set array', () => {
    const alerts = [{}];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
  
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if prop alerts set' +
    ' [{vertical:false,buttons:[],content:"test"}]', () => {
    const alerts = [{
      vertical: false,
      buttons: [],
      content: 'test',
    }];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
  
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if prop alerts set' +
    ' [{vertical:true,buttons:[],content:""}]', () => {
    const alerts = [{
      vertical: true,
      buttons: [],
      content: 'test',
    }];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if prop alerts has buttons and checks them', () => {
    const alerts = [{
      vertical: true,
      buttons: [{
        text: 'test',
        action: () => {},
        closeAlert: true,
      }],
      content: 'test',
    }];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
  
    expect(tree.toJSON()).toMatchSnapshot();
    findByType(tree.toJSON(), 'button').props.onClick();
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if prop alerts set several vertical items', () => {
    const alerts = [
      {
        vertical: true,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 1',
      },
      {
        vertical: true,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 2',
      },
    ];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
  
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if prop alerts set several vertical/horizontal' +
    ' items', () => {
    const alerts = [
      {
        vertical: false,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 1',
      },
      {
        vertical: true,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 2',
      },
    ];
  
    const tree = renderer.create(
      <AlertArea alerts={alerts} />,
    );
  
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('handles prop change', () => {
    const alerts = [
      {
        vertical: false,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 1',
      },
      {
        vertical: true,
        buttons: [{
          text: 'close',
          action: () => {},
          closeAlert: true,
        }],
        content: 'TEST 2',
      },
    ];
  
    const component = mount(
      <AlertArea alerts={alerts} />,
    );
    component.setProps({ alerts: [] });
    expect(toJson(component)).toMatchSnapshot();
  });

  it('ensures that component methods are working', () => {
    const tree = renderer.create(
      <AlertArea />,
    );
    const instanse = tree.getInstance();
    instanse._handleAlertClose();
    instanse.timeout = null;
    instanse.addToQueue({
      vertical: false,
      buttons: [{
        text: 'close',
        action: () => {},
        closeAlert: true,
      }],
      content: 'TEST',
    });
    instanse._closeCurrentAlert();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
