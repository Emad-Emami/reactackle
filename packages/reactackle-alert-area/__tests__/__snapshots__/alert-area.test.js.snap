// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertArea/> ensures that component methods are working 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
/>
`;

exports[`<AlertArea/> handles prop change 1`] = `
<AlertArea
  alerts={Array []}
  animationDuration={300}
  defaultTimeout={3000}
  onClose={[Function]}
  positionX="right"
  positionY="bottom"
>
  <AlertsWrapperStyled
    positionX="right"
    positionY="bottom"
  >
    <div
      className="AlertsWrapperStyled-iQzJHQ eedxzK"
    />
  </AlertsWrapperStyled>
</AlertArea>
`;

exports[`<AlertArea/> renders correctly if prop alerts has buttons and checks them 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM cadAdM"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      test
    </div>
    <div
      className="ActionsStyled-hhpOda dAXLwA"
    >
      <div
        className="ActionsWrapperStyled-eCnBHy buEQNK"
      >
        <button
          className="ButtonStyled__Button-lkXKpB iDqeCe"
          disabled={false}
          onClick={[Function]}
          radius="default"
          size="normal"
          title=""
        >
          <div
            className="ButtonTextStyled-abyTl aDQbB"
            size="normal"
          >
            <div
              className="ButtonTitleStyled-idIFzn gXbrKW"
              size="normal"
            >
              test
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts has buttons and checks them 2`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM iZMNkH"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      test
    </div>
    <div
      className="ActionsStyled-hhpOda dAXLwA"
    >
      <div
        className="ActionsWrapperStyled-eCnBHy buEQNK"
      >
        <button
          className="ButtonStyled__Button-lkXKpB iDqeCe"
          disabled={false}
          onClick={[Function]}
          radius="default"
          size="normal"
          title=""
        >
          <div
            className="ButtonTextStyled-abyTl aDQbB"
            size="normal"
          >
            <div
              className="ButtonTitleStyled-idIFzn gXbrKW"
              size="normal"
            >
              test
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts set [{vertical:false,buttons:[],content:"test"}] 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM iqNaVU"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      test
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts set [{vertical:true,buttons:[],content:""}] 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM cadAdM"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      test
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts set array 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM iqNaVU"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    />
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts set several vertical items 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM cadAdM"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      TEST 1
    </div>
    <div
      className="ActionsStyled-hhpOda dAXLwA"
    >
      <div
        className="ActionsWrapperStyled-eCnBHy buEQNK"
      >
        <button
          className="ButtonStyled__Button-lkXKpB iDqeCe"
          disabled={false}
          onClick={[Function]}
          radius="default"
          size="normal"
          title=""
        >
          <div
            className="ButtonTextStyled-abyTl aDQbB"
            size="normal"
          >
            <div
              className="ButtonTitleStyled-idIFzn gXbrKW"
              size="normal"
            >
              close
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly if prop alerts set several vertical/horizontal items 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
>
  <div
    className="AlertStyled-tHNAM iqNaVU"
    role="alert"
  >
    <div
      className="ContentStyled-dbeIsI clUWdm"
    >
      TEST 1
    </div>
    <div
      className="ActionsStyled-hhpOda dAXLwA"
    >
      <div
        className="ActionsWrapperStyled-eCnBHy buEQNK"
      >
        <button
          className="ButtonStyled__Button-lkXKpB iDqeCe"
          disabled={false}
          onClick={[Function]}
          radius="default"
          size="normal"
          title=""
        >
          <div
            className="ButtonTextStyled-abyTl aDQbB"
            size="normal"
          >
            <div
              className="ButtonTitleStyled-idIFzn gXbrKW"
              size="normal"
            >
              close
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AlertArea/> renders correctly with default props 1`] = `
<div
  className="AlertsWrapperStyled-iQzJHQ eedxzK"
/>
`;
